### PR TITLE
 RenderImage: factorize all image rendering and scaling code 

### PR DESCRIPTION
**Pull request by poire-z**
RenderImage: factorize all image rendering and scaling code:
New module RenderImage (alongside existing RenderText) to provide image rendering and scaling facilities.
Uses MuPDF, but tries first giflib on GIF. Allows for getting all the frames from an animated GIF.
Needs a base bump for https://github.com/koreader/koreader-base/pull/657

PicDocument:getCoverPageImage(): use first page instead of MuPDF:
The dedicated image format library may often give better results than MuPDF (eg: for animated GIFs).

ImageViewer: allow viewing a list of images:
Mostly useful for browsing the frames of animated GIFs (but generic enough for other uses)
It adds a small progress bar under image, and allow switching frame with Tap on left/right of screen:
<kbd>image</kbd>

### CODE REVIEW

**Comment by poire-z** [comment]
Note: I removed the following code, as I checked that with our current (updated since the time I added this) MuPDF does no more `free()` the userdata provided in case of error:
https://github.com/koreader/koreader/blob/d601eabc4ad516f07266e6a809c3322b8b83933c/frontend/document/credocument.lua#L207-L216

**Comment by onde2rock** [comment]
Nice, I did not test but it doesn't seem to be in there, could you had something like : 

```
    if Device:hasKeys() then
        self.key_events.NextPage = {
            {Input.group.PgFwd}, doc = "goto next page"
        }
        self.key_events.PrevPage = {
            {Input.group.PgBack}, doc = "goto previous page"
        }

```

**Comment by poire-z** [comment]
It currently has:
https://github.com/koreader/koreader/blob/d601eabc4ad516f07266e6a809c3322b8b83933c/frontend/ui/widget/imageviewer.lua#L72-L79
because at the time, there was no prev/next page as it _may_ have now.
Cause 99% of the images we may find in epub or cover are not animated (sidenote: I found a very little number of GIF, all non animated) among my epubs, they are mostly made of PNG or JPEG).

Should I make PgFwd/PgBk do next/prev only if animated GIF, and zoom+/- otherwise (inconsistent behaviour?).

**Comment by onde2rock** [comment]
```
             ZoomIn = { {Device.input.group.PgBack}, doc = "Zoom In" }, 
             ZoomOut = { {Device.input.group.PgFwd}, doc = "Zoom out" }, 
```

Neat, I didn't know that was a thing.


<br>

> Should I make PgFwd/PgBk do next/prev only if animated GIF, and zoom+/- otherwise (inconsistent behaviour?).

Since there is no way to move the zoomed in picture, and this is a very little known feature, I'd say, do nothing. Don't worry about it.

**Comment by poire-z** [comment]
OK (although on the emulator, while testing it, I too often wanted to use PgUp/PgDown to change frames, and the zoom was bothering).
Also, this is mostly useful with Wikipedia epub, and there are few chances such animated GIF are the cover image (that you could reach thru menu). They would mostly be shown in the EPUB, and you would need to Hold on the image displayed by crengine to launch ImageViewer. I guess that without a touch screen, you can't do that on your Kindle non-touch :|

**Comment by onde2rock** [comment]
> (although on the emulator, while testing it, I too often wanted to use PgUp/PgDown to change frames, and the zoom was bothering

Or if you want to add it, you can just remove the zoom key_events, it's useless.

 PgUp/PgDown to change frames is more intuitive anyway.

**Comment by poire-z** [comment]
No worry about that, the various getCoverPageImage() would all get only the first frame if the cover image would be animated (and I guess there is probably no epub in the world with an animated gif as cover :) so I won't bother making that happen :)

The only places where an explicit request for all the animated gif frames (if there are) is made is:
- in crengine, when holding on an embedded image, to display it with ImageViewer
- in wikipedia lookup results, when holding on an image (where there are often interesting ones, which would also get saved and available in the EPUB)

**Comment by Frenzie** [comment]
http://www.creativindiecovers.com/how-to-make-an-animated-book-cover-that-blows-peoples-minds/

**Comment by KenMaltby** [comment]
That could catch on, especially if they work on the web stores.  So it might be useful someday.

**Comment by poire-z** [comment]
Semi-random PR hijack to mention something strange, pinging @NiLuJe .

When trying to use RenderImage:scaleBlitBuffer() in #8426 to have nicer scaling, this worked until I tried how it look in landscape mode, on a Kobo or in the emulator:

```lua
        local bb = Screen.bb:copy() -- this makes MuPDF crashes when in landscape mode
        -- local bb = Blitbuffer.new(Screen:getWidth(), Screen:getHeight(), Screen.bb:getType()) -- this doesn't, and it's what I'll used as it is cheaper than copy()
        -- bb:scale(thumbsize.w, thumbsize.h) -- this was ok with :copy()
        local RenderImage = require("ui/renderimage")
        bb = RenderImage:scaleBlitBuffer(bb, thumbsize.w, thumbsize.h, true),
```

MuPDF complains about **Illegal stride for pixmap** : 

```
12/09/21-12:36:21 DEBUG RenderImage:scaleBlitBuffer: scaling
error: Illegal stride for pixmap (n=4 w=610, stride=2400)
error in subprocess: ./ffi/mupdf.lua:78: could not create pixmap from blitbuffer: Illegal stride for pixmap (n=4 w=610, stride=2400) (2)
stack traceback:
   [C]: in function 'error'
   ./ffi/mupdf.lua:78: in function 'merror'
   ./ffi/mupdf.lua:887: in function 'scaleBlitBuffer'
   frontend/ui/renderimage.lua:185: in function 'scaleBlitBuffer'
   frontend/ui/widget/bookmapwidget.lua:1395: in function 'func'
   ./ffi/util.lua:343: in function <./ffi/util.lua:314>
   [C]: in function 'xpcall'
   ./ffi/util.lua:314: in function 'runInSubProcess'
   frontend/ui/widget/bookmapwidget.lua:1363: in function 'startTileGeneration'
   frontend/ui/widget/bookmapwidget.lua:1351: in function 'ensureTileGeneration'
   frontend/ui/widget/bookmapwidget.lua:1332: in function 'showPageThumbnail'
   frontend/ui/widget/bookmapwidget.lua:1206: in function 'handleEvent'
   frontend/ui/widget/container/inputcontainer.lua:263: in function 'handleEvent'
   frontend/ui/uimanager.lua:1119: in function 'sendEvent'
   frontend/ui/uimanager.lua:53: in function '__default__'
   frontend/ui/uimanager.lua:1603: in function 'handleInputEvent'
   frontend/ui/uimanager.lua:1683: in function 'handleInput'
   frontend/ui/uimanager.lua:1730: in function 'run'
   ./reader.lua:290: in main chunk
   [C]: at 0x559252dfa7b1
```
Earlier on my Kobo, I got:
```
  error: Illegal stride for pixmap (n=1 w=1448, stride=1088)
  error in subprocess:    ./ffi/mupdf.lua:78: could not create pixmap from blitbuffer: Illegal stride for pixmap (n=1 w=1448, stride=1088) (2)
```

So, it's possible there is something missing/fishy in blitbuffer.lua in :copy(), or in framebuffer.lua when doing the rotation - but all hit is way over my head :/
No issue for me as I can just use BB.new instead of :copy - but well, thought you might be interested in this possible bug.

**Comment by poire-z** [comment]
For info, I've been benchmarking our various ways of scaling down a blitbuffer.
Scaling a blitbuffer of 600x610 on the emulator - or portrait on my Kobo, to have its size half of the original:

`bb = bb:scale()` (nearest neighbour)

`RenderImage:scaleBlitBuffer(bb, thumbsize.w, thumbsize.h, true)` MuPDF scaling

And via your QT smoothscaling from crengine:
```c
static int smoothScaleBlitBuffer(lua_State *L) {
    BlitBuffer *bb = (BlitBuffer*) lua_topointer(L, 1);
    int dw = luaL_checkint(L, 2);
    int dh = luaL_checkint(L, 3);
    lUInt8* data = CRe::qSmoothScaleImage((const lUInt8*)bb->data, bb->w, bb->h, true, dw, dh);
    if (!data)
        return 0;
    lua_pushlightuserdata(L, (void*)data);
    lua_pushinteger(L, dw*dh*4);
    return 2;
}
```
and
```lua
        local data, size = cre.smoothScaleBlitBuffer(bb, thumbsize.w, thumbsize.h)
        local scaled_bb = Blitbuffer.new(thumbsize.w, thumbsize.h, Blitbuffer.TYPE_BBRGB32, data)
        scaled_bb:setAllocated(1) -- need to free data malloc'ed by cre
        if Screen.bb:getBpp() <= 8 then
            logger.warn(string.format("  intermediate scaling took %.3f seconds", TimeVal:getDuration(start_tv)))
            bb:free() -- reclaim this memory so we don't add
            local bb8 = Blitbuffer.new(thumbsize.w, thumbsize.h, TYPE_BB8)
            bb8:blitFrom(scaled_bb)
            scaled_bb = bb8
        end
```

Results:

how | Emulator x64 | Kobo arm
------ | ----------- | ---------------
`bb:scale()` | 11ms | 140 ms
`Mupdf.scaleBlitBuffer()` | 15ms | 40ms
`cre.smoothScaleBlitBuffer()` | 3ms (!) (+1ms if I force convert from 32 to 8) | 50ms (+15ms to convert from 32bpp to 8bpp)

Strange that cre.smoothScaleBlitBuffer is so much faster on x64, but not at all on arm.
Anyway, I'll go with MuPDF as it's just there and the standard for that.

**Comment by Frenzie** [comment]
@poire-z Which compiler did you use for Kobo arm btw? If it's the standard OS armhf cross-compiler results might be comparatively suboptimal, although I shouldn't normally expect anything that drastic.

**Comment by poire-z** [comment]
arm-linux-gnueabihf-gcc (Debian 8.3.0-2) 8.3.0 - from Debian 10
vs.
emulator gcc (Debian 10.2.1-6) 10.2.1 20210110 - from Debian 11

**Comment by Frenzie** [comment]
> arm-linux-gnueabihf-gcc (Debian 8.3.0-2) 8.3.0 - from Debian 10

How much is the question, but that should in principle be worse than our own TC (which I compile as part of creating the koreader/kokobo Docker image). We'll have some CPU/FPU features enabled that a more generic armhf TC might not,[1] and of course we use GCC 11 which should normally be better too.

I think it might be worthwhile to check against the binaries for all the libraries and LuaJIT from our latest nightly instead. (Heck, if the outcome is that it barely matters, that's good to know too! ;-) )

For the most part see https://github.com/NiLuJe/crosstool-ng/blob/3e2d99d030cbecc0119aa08bd06d062f59669d45/samples/arm-kobo-linux-gnueabihf/crosstool.config for details.

[1] And the feature under discussion does sound like the kind of thing that might involve floating point computation?

**Comment by Frenzie** [comment]
In that case I doubt it matters as long as libcrengine.so is from the build server, though I suppose it'd be cleanest if they were all the same.

**Comment by poire-z** [comment]
Well, I have to build libcrengine.so too, otherwise my libkoreader-cre.so crashes when used with a nightly libcrengine.so.
But if you build me a likoreader-cre.so with the same tools used by the nightly, it might work with a nightly libcrengine.so (but you can zip them both, in case I need a compatible libcrengine.so)

**Comment by Frenzie** [comment]
Release or debug?

**Comment by poire-z** [comment]
Release :) if we are benchmarking. Same as how the nightlies are (as that's their mupdf I'm benchmarking against).

**Comment by Frenzie** [comment]
I think this should be it.
[libs.zip](https://github.com/koreader/koreader/files/7686252/libs.zip)

**Comment by poire-z** [comment]
Thanks (won't be able to test them until this evening).

**Comment by poire-z** [comment]
^tested, no change, still 50ms (+15ms to convert from 32bpp to 8bpp):
```
12/09/21-19:38:27 WARN    intermediate scaling took 0.055 seconds
12/09/21-19:38:27 WARN    scaling took 0.072 seconds, 8 bpp
12/09/21-19:38:27 WARN    intermediate scaling took 0.050 seconds
12/09/21-19:38:27 WARN    scaling took 0.066 seconds, 8 bpp
12/09/21-19:38:28 WARN    intermediate scaling took 0.053 seconds
12/09/21-19:38:28 WARN    scaling took 0.069 seconds, 8 bpp
```

**Comment by NiLuJe** [comment]
The Qt code has better assembly support for x64 than arm ;).

**Comment by Frenzie** [comment]
> ^tested, no change, still 50ms (+15ms to convert from 32bpp to 8bpp):

Thanks, makes sense, but good to know. :-)

**Comment by NiLuJe** [comment]
On mobile, but copy looks sane, I'll try to see if I can repro and see
what's upsetting mupdf this we ;).

**Comment by NiLuJe** [comment]
Okay, that's because we never physically rotate buffers, we rotate during blit.

TL;DR: `RenderImage:scaleBlitBuffer` should probably blit to a temp buffer and scale *that* if the source buffer is rotated.

**Comment by poire-z** [comment]
> should probably do a simple blit to a temp buffer 

I'd rather avoid having to create another (huge) buffer. No way before invoking MuPDF to (temporarily) hack the bb and update the stride stuff so it matches what it expects?

**Comment by NiLuJe** [comment]
Nope, because it's not actually rotated ^^.

You'd be scaling a thing in Portrait and then blitting over a Landscape layout :?

**Comment by NiLuJe** [comment]
Or.. is it?

Try just swapping w & h in your MuPDF call?

**Comment by poire-z** [comment]
(I ended up not using :copy and create a new buffer, so I'm no longer affected. So, we're just talking future usages :)
Or we could just notice it's a fake rotated, switche the target_w and target_h, scale it to that, and flag the resulting bb to it gets the same rotation flags as the original ?

**Comment by NiLuJe** [comment]
> Or we could just notice it's a fake rotated, switche the target_w and target_h, scale it to that, and flag the resulting bb to it gets the same rotation flags as the original ?

Probably, that's what I was thinking of in the previous reply ;).

**Comment by poire-z** [comment]
Something like that ?
```diff
 function RenderImage:scaleBlitBuffer(bb, width, height, free_orig_bb)
     if not width or not height then
         logger.dbg("RenderImage:scaleBlitBuffer: no need")
         return bb
     end
     -- Ensure we give integer width and height to MuPDF, to
     -- avoid a black 1-pixel line at right and bottom of image
     width, height = math.floor(width), math.floor(height)
     if bb:getWidth() == width and bb:getHeight() == height then
         logger.dbg("RenderImage:scaleBlitBuffer: no need")
         return bb
     end
     logger.dbg("RenderImage:scaleBlitBuffer: scaling")
     local scaled_bb
     if G_reader_settings:isTrue("legacy_image_scaling") then
         -- Uses "simple nearest neighbour scaling"
         scaled_bb = bb:scale(width, height)
     else
+        local bb_rotation_mode = bb:getRotation()
+        if bb_rotation_mode % 2 == 1 then
+            width, height = height, width
+        end
+        bb:setRotation(0)
         -- Better quality scaling with MuPDF
         if not Mupdf then Mupdf = require("ffi/mupdf") end
         scaled_bb = Mupdf.scaleBlitBuffer(bb, width, height)
+        bb:setRotation(bb_rotation_mode)
+        scaled_bb:setRotation(bb_rotation_mode) -- XXX
     end
     if not free_orig_bb == false then
         bb:free()
     end
     return scaled_bb
 end
```
It does not crash anymore.
But whether the line marked XXX is in or out, I get:
image

I'm sure you'll see what is wrong and avoid me from having to understand all that :)

**Comment by NiLuJe** [comment]
It's unclear to me what's the source, what's the target, and how either are rendered in the first place ;o).

Is that code up somewhere?

**Comment by NiLuJe** [comment]
(Possibly, the final setRotation ought to be to the *target's* mode (e.g., Screen?), not the source? (Depending on how the source is rendered, obviously).

**Comment by NiLuJe** [comment]
And/or just "fix" MuPDF to ignore soft rotations:

```
diff --git a/ffi/mupdf.lua b/ffi/mupdf.lua
index 9322d4be..203be80b 100644
--- a/ffi/mupdf.lua
+++ b/ffi/mupdf.lua
@@ -832,7 +832,6 @@ Same arguments as BlitBuffer:scale() for easy replacement.
 --]]
 function mupdf.scaleBlitBuffer(bb, width, height)
     -- We need first to convert our BlitBuffer to a pixmap
-    local orig_w, orig_h = bb:getWidth(), bb:getHeight()
     local bbtype = bb:getType()
     local colorspace
     local converted_bb
@@ -862,15 +861,15 @@ function mupdf.scaleBlitBuffer(bb, width, height)
         end
         alpha = 1
     elseif bbtype == BlitBuffer.TYPE_BB4 then
-        converted_bb = BlitBuffer.new(orig_w, orig_h, BlitBuffer.TYPE_BB8)
-        converted_bb:blitFrom(bb, 0, 0, 0, 0, orig_w, orig_h)
+        converted_bb = BlitBuffer.new(bb.w, bb.h, BlitBuffer.TYPE_BB8)
+        converted_bb:blitFrom(bb, 0, 0, 0, 0, bb.w, bb.h)
         bb = converted_bb -- we don't free() the provided bb, but we'll have to free our converted_bb
         colorspace = M.fz_device_gray(context())
         alpha = 0
-        stride = orig_w
+        stride = bb.w
     else
-        converted_bb = BlitBuffer.new(orig_w, orig_h, BlitBuffer.TYPE_BBRGB32)
-        converted_bb:blitFrom(bb, 0, 0, 0, 0, orig_w, orig_h)
+        converted_bb = BlitBuffer.new(bb.w, bb.h, BlitBuffer.TYPE_BBRGB32)
+        converted_bb:blitFrom(bb, 0, 0, 0, 0, bb.w, bb.h)
         bb = converted_bb -- we don't free() the provided bb, but we'll have to free our converted_bb
         if mupdf.bgr then
             colorspace = M.fz_device_bgr(context())
@@ -881,7 +880,7 @@ function mupdf.scaleBlitBuffer(bb, width, height)
     end
     -- We can now create a pixmap from this bb of correct type
     local pixmap = W.mupdf_new_pixmap_with_data(context(), colorspace,
-                    orig_w, orig_h, nil, alpha, stride, ffi.cast("unsigned char*", bb.data))
+                    bb.w, bb.h, nil, alpha, stride, ffi.cast("unsigned char*", bb.data))
     if pixmap == nil then
         if converted_bb then converted_bb:free() end -- free our home made bb
         merror("could not create pixmap from blitbuffer")
```

**Comment by NiLuJe** [comment]
c.f., also `ui/widget/imagewidget` use of `bb.rotatedCopy`?

**Comment by poire-z** [comment]
Some earlier version (too much a mess now to update it, althought I switched from dump to ffi.buffer.encode via Persist.lua) is at: https://github.com/poire-z/koreader/blob/e2b2f99b3b7e99dfa0a2419cbc334234e46910d9/frontend/ui/widget/bookmapwidget.lua#L1291-L1309

One thing to note is that I use TileCacheItem:totable/fromtable() to serialize it and pass it back from subprocess.
Is the rotation flag preserved in this? : 
https://github.com/koreader/koreader/blob/5ed22cafb5f18aa5a9200a77387669198dd39f78/frontend/document/tilecacheitem.lua#L15-L29

**Comment by poire-z** [comment]
Quickly/ugly putting the code of the subprocess in the parent process:
- your mupdf.lua patch doesn't help with the rotation (not did it when in subprocess), but it avoids the illegal stride error (may be it was just aimed at that :)
- my renderimage.lua setRotation and w/h switch stuff does help with getting the correct rotation when bliting.

So, I guess:
- your mudpf patch is good (it's just cleaning this orig_w/h stuff, did this really have no purpose?)
- rotation is lost when in TileCacheItem and serialized/unserialized: dunno how much this is an issue (with PDF documents in landscape and re-opening and using cached tiles?)
- don't know how much my renderimage.lua really matters in other contexts (but I don't even need it in my context if I don't use `bb = Screen.bb:copy()`)

**Comment by NiLuJe** [comment]
>     * your mudpf patch is good (it's just cleaning this orig_w/h stuff, did this really have no purpose?)

Yeah, it made sense before I wrote the fast zero-copy paths, because the blitFrom fallback was then always used.

**Comment by NiLuJe** [comment]
> Quickly/ugly putting the code of the subprocess in the parent process:
> 
>     * your mupdf.lua patch doesn't help with the rotation (nor did it when in subprocess), but it avoids the illegal stride error (may be it was just aimed at that :)
> 
>     * my renderimage.lua setRotation and w/h switch stuff does help with getting the correct rotation when bliting.
> 
> 
> So, I guess:
> 
>     * your mudpf patch is good (it's just cleaning this orig_w/h stuff, did this really have no purpose?)
> 
>     * rotation is lost when in TileCacheItem and serialized/unserialized: dunno how much this is an issue (with PDF documents in landscape and re-opening and using cached tiles?)
> 

It is, yeah. Dinner now, but I can probably whip up a quick fix tonight ;).

**Comment by NiLuJe** [comment]
>     * your mupdf.lua patch doesn't help with the rotation (nor did it when in subprocess), but it avoids the illegal stride error (may be it was just aimed at that :)

Yeah, that was the intent ;). (i.e., just import the bb as-is, memory layout wise, which ignores rotation).

**Comment by NiLuJe** [comment]
This should ensure TileCache actually ser/deser a BB properly ;).

```diff
diff --git a/ffi/blitbuffer.lua b/ffi/blitbuffer.lua
index 997c5348..4a57cc2b 100644
--- a/ffi/blitbuffer.lua
+++ b/ffi/blitbuffer.lua
@@ -2148,11 +2148,17 @@ function BB.compat(oldbuffer)
     return ffi.cast("BlitBuffer4*", oldbuffer)[0]
 end
 
-function BB.fromstring(width, height, buffertype, str, stride)
+function BB.fromstring(width, height, buffertype, str, stride, rotation, inverse)
     local dataptr = C.malloc(#str)
     ffi.copy(dataptr, str, #str)
     local bb = BB.new(width, height, buffertype, dataptr, stride)
     bb:setAllocated(1)
+    if rotation ~= nil then
+        bb:setRotation(rotation)
+    end
+    if inverse ~= nil then
+        bb:setInverse(inverse)
+    end
     return bb
 end
```

```diffdiff --git a/frontend/document/tilecacheitem.lua b/frontend/document/tilecacheitem.lua
index 60b4b621..6ac9cef0 100644
--- a/frontend/document/tilecacheitem.lua
+++ b/frontend/document/tilecacheitem.lua
@@ -24,6 +24,8 @@ function TileCacheItem:totable()
             h = self.bb.h,
             stride = tonumber(self.bb.stride),
             fmt = self.bb:getType(),
+            rotation = self.bb:getRotation(),
+            inverse = self.bb:getInverse(),
             data = Blitbuffer.tostring(self.bb),
         },
     }
@@ -54,7 +56,7 @@ function TileCacheItem:fromtable(t)
     self.excerpt = t.excerpt
     self.created_ts = t.created_ts
     self.persistent = t.persistent
-    self.bb = Blitbuffer.fromstring(t.bb.w, t.bb.h, t.bb.fmt, t.bb.data, t.bb.stride)
+    self.bb = Blitbuffer.fromstring(t.bb.w, t.bb.h, t.bb.fmt, t.bb.data, t.bb.stride, t.bb.rotation, t.bb.inverse)
 end
 
 function TileCacheItem:load(filename)
```

**Comment by poire-z** [comment]
^confirmed, that makes the job done in subprocess works as when I put it in the parent process.
https://github.com/koreader/koreader/pull/3905#issuecomment-991709144 is still needed to solve my (initial) rotation issue - dunno if you want to tackle this case in any (other?) way.
Anyway, all this is not strictly related to my current experiments, so feel free to PR all this :)

**Comment by NiLuJe** [comment]
> ^confirmed, that makes the job done in subprocess works as when I put it in the parent process. [#3905 (comment)](https://github.com/koreader/koreader/pull/3905#issuecomment-991709144) is still needed to solve my (initial) rotation issue - dunno if you want to tackle this case in any (other?) way. Anyway, all this is not strictly related to my current experiments, so feel free to PR all this :)

I'm torn. I've thought about putting it inside `mupdf.scaleBlitBuffer`, but I kind of like it out, as it ensure the resulting buffer is *always* unrotated, which is (much) better for performance. And we're probably using a few of those in somewhat latency critical code.

(It's kind of the same reasoning as the ImageWidget rotatedCopy stuff: while we could probably handle it in a slightly similar fashion, using `rotatedCopy` ensures we get an unrotated buffer in the end).

**Comment by NiLuJe** [comment]
> ^confirmed, that makes the job done in subprocess works as when I put it in the parent process. [#3905 (comment)](https://github.com/koreader/koreader/pull/3905#issuecomment-991709144) is still needed to solve my (initial) rotation issue - dunno if you want to tackle this case in any (other?) way. Anyway, all this is not strictly related to my current experiments, so feel free to PR all this :)

Besides hiding all that under a `if source_bb_rota ~= 0 and bit.band(1, source_bb:getRotation()) == 1 then` branch, I mean ;p.

(`&` is faster than `%` here ;)).

**Comment by poire-z** [comment]
> (It's kind of the same reasoning as the ImageWidget rotatedCopy stuff: while we could probably handle it in a slightly similar fashion, using rotatedCopy ensures we get an unrotated buffer in the end).

But in ImageWidget, we may need a copy if it ends up being cached - or at least, it's what is returned/displayed and have a life of its own after that.
In RenderImage:scaleBlitBuffer(), it would be the intermediate/temporary BB, so any trick feels better than having to allocate a copy of the input BB (which might be as huge as the Screen BB :)

**Comment by NiLuJe** [comment]
Yup, that was exactly what I was (failing, apparently? :D) to get across ;p.
